lgc/patch/MeshTaskShader.cpp

```diff
@@ -541,4 +1054,1365 @@ Value *MeshTaskShader::convertToDivergent(Value *value) {
+      readValue = m_builder->CreateBitCast(readValue, readTy);
+
+    return readValue;
+  }
+
+  readPtr = m_builder->CreateBitCast(readPtr, PointerType::get(readTy, readPtr->getType()->getPointerAddressSpace()));
+  return m_builder->CreateAlignedLoad(readTy, readPtr, Align(4));
+}
+
+// =====================================================================================================================
+// Write value to mesh shader LDS.
+//
+// @param writeValue : Value to write
+// @param ldsOffset : LDS offset in dwords
+void MeshTaskShader::writeValueToLds(Value *writeValue, Value *ldsOffset) {
+  assert(m_lds);
+
+  auto writeTy = writeValue->getType();
+  assert(writeTy->isIntOrIntVectorTy() || writeTy->isFPOrFPVectorTy());
+
+  Value *writePtr = m_builder->CreateGEP(m_lds->getValueType(), m_lds, {m_builder->getInt32(0), ldsOffset});
+
+  const unsigned bitWidth = writeTy->getScalarSizeInBits();
+  if (bitWidth == 8 || bitWidth == 16) {
+    // For 8-bit and 16-bit data type, we write them as 32-bit types to LDS. They are not packed tightly in LDS.
+    unsigned numElems = writeTy->isVectorTy() ? cast<FixedVectorType>(writeTy)->getNumElements() : 1;
+
+    if (writeTy->isFPOrFPVectorTy()) {
+      Type *castTy = m_builder->getIntNTy(bitWidth);
+      if (writeTy->isVectorTy())
+        castTy = FixedVectorType::get(m_builder->getIntNTy(bitWidth), numElems);
+
+      writeValue = m_builder->CreateBitCast(writeValue, castTy);
+    }
+
+    Type *extTy = m_builder->getInt32Ty();
+    if (writeTy->isVectorTy())
+      extTy = FixedVectorType::get(m_builder->getInt32Ty(), numElems);
+
+    writeValue = m_builder->CreateZExt(writeValue, extTy);
+
+    writePtr = m_builder->CreateBitCast(
+        writePtr, PointerType::get(writeValue->getType(), writePtr->getType()->getPointerAddressSpace()));
+    m_builder->CreateAlignedStore(writeValue, writePtr, Align(4));
+    return;
+  }
+
+  writePtr = m_builder->CreateBitCast(
+      writePtr, PointerType::get(writeValue->getType(), writePtr->getType()->getPointerAddressSpace()));
+  m_builder->CreateAlignedStore(writeValue, writePtr, Align(4));
+}
+
+// =====================================================================================================================
+// Do atomic binary operation with the value stored in LDS.
+//
+// @param atomicOp : Atomic binary operation
+// @param atomicValue : Value to do atomic operation
+// @param ldsOffset : LDS offset in dwords
+void MeshTaskShader::atomicOpWithLds(AtomicRMWInst::BinOp atomicOp, Value *atomicValue, Value *ldsOffset) {
+  assert(atomicValue->getType()->isIntegerTy(32));
+
+  Value *atomicPtr = m_builder->CreateGEP(m_lds->getValueType(), m_lds, {m_builder->getInt32(0), ldsOffset});
+  auto atomicInst = m_builder->CreateAtomicRMW(atomicOp, atomicPtr, atomicValue, MaybeAlign(),
+                                               AtomicOrdering::SequentiallyConsistent, SyncScope::System);
+  atomicInst->setVolatile(true);
```

Comment:
The way this method is used, the atomics are not used to synchronize between threads. They're used as an optimization to avoid using the VALU for read-modify-write. Because of that, the atomic doesn't have to be volatile, and it can use `AtomicOrdering::Monotonic` and `SyncScope::Singlethread`. This should help the backend understand that no waitcount instructions are needed around these atomics.

Reply:
Thank you. I change the semantics. And backend doesn't insert s_waitcnt as expected. I see HW has ds_mskor instruction.  But backend doesn't expose it. So we still switch to use LDS atomics to achieve the same thing.